Create document on safely executing repros Fixes #45341 (#45456)
dotnet/sdk: documentation/general/ExecutingCustomerCode.md

# Executing Customer Code

## Summary

When customers discover issues with the .NET SDK, we often need more information to see how exactly their scenarios differ from the (presumably working) mainline scenarios. This additional information often takes the form of a 'repro' or set of steps by which we can see the error ourselves and walk through what is happening as their scenario plays out and how it ultimately diverged from our expectations.

> Blindly executing unvetted customer code can be a security hazard, however, not just for the machine executing the code but for any machine on the same network and any service accessible using credentials they can access through those machines. In this way, a malicious actor can exfiltrate sensitive Microsoft data, including information about other Microsoft employees, proprietary code, or private customer data. They may even be able to take down a service or introduce further security bugs in shipping products. Indeed, the most common vector hackers use to gain access is through compromising one or more individual users with employee credentials. At Microsoft where security is paramount, we want to prevent such hacks.

Reviewer: ```suggestion
Blindly executing, or even just opening in IDE without intention to build, unvetted customer code can be a security hazard, however, not just for the machine executing the code but for any machine on the same network and any service accessible using credentials they can access through those machines. In this way, a malicious actor can exfiltrate sensitive Microsoft data, including information about other Microsoft employees, proprietary code, or private customer data. They may even be able to take down a service or introduce further security bugs in shipping products. Indeed, the most common vector hackers use to gain access is through compromising one or more individual users with employee credentials. At Microsoft where security is paramount, we want to prevent such hacks.
```

> This document contains recommended best practices on how to securely test users' code. They are arranged in order or security with the most secure at the top. This should also be the priority you should use to stay secure when executing code.

Reviewer: ```suggestion
This document contains recommended best practices on how to securely test users' code. They are arranged in order of security with the most secure at the top. This should also be the priority you should use to stay secure when executing code.
```

## Strategies for Staying Secure

### Binary Logs

MSBuild binary logs (binlogs) are structured data covering what happened during a build. Many operations that aren't obviously 'builds' such as publish (`dotnet publish`) or pack (`dotnet pack`) include building as part or all of the process, which makes them very useful for learning more about and potentially diagnosing a variety of potential problems. Furthermore, they are safe in the sense that analyzing a binlog does not expose your system to risks.

To provide information to customers about collecting binlogs, direct them to:
https://aka.ms/msbuild/binlog

> It is important to note, however, that binlogs often collect secrets as part of the information relevant to the build. A binlog of a simplified repro can work as well, but it may be difficult to create a simple repro. Customers are sometimes concerned (with good reason) about sharing binlogs of their real (or simplified) builds. If that is the case, or if the reported issue does not involve MSBuild, another option may be more relevant.

Reviewer: Note there are currently options for redact secrets from the binlog - either via viewer (https://msbuildlog.com/#redaction) or with binlogtool (`binlogtool redact`)

3. Sandbox is essentially empty when you first create it. It has internet access as long as your host computer has internet access, but it will be missing various programs and tools you may need to reproduce the customer's problem. Such programs and tools need to be downloaded anew every time you want to use Sandbox.

Conveniently, however, it is relatively easy to copy files from your host computer into Sandbox, and as noted earlier, it is a secure, isolated environment, so it is an excellent choice if it is viable.

### Create Your Own Solution

Depending on how well you understand the customer's scenario, you may or may not understand ahead of time the most likely problem, but if it's possible to create your own solution that reproduces the issue the customer alluded to in their report, and you understand all the steps it took to create that solution, that is a safe method for obtaining a project to test and debug. Note that this is only a secure method if you understand all steps involved in creating the test project. "Copy project file provided by customer to disk and build it" does not count as a well-understood pair of steps.

> ### Create a Virtual Machine Without Your Credentials

Reviewer: Should we add an option for running within docker?
There area readily available images from various channels of sdk (nightly/preview/release) https://hub.docker.com/r/microsoft/dotnet 

It feels more lightweight then full blown VM

Creating and using a virtual machine (VM) without your credentials is another secure way to test a customer's code, as it's disconnected from either your credentials or your computer. It is important if using this method to ensure that you do not give your VM your credentials at any point such as, for example, signing into your Microsoft account so you can use Visual Studio Enterprise.

The other disadvantage of this approach is that it can take a lot of time to select appropriate parameters for your VM, start it, connect to it, set it up properly for your scenario, and execute the scenario. Unlike Sandbox, it can retain certain information across restarts, but it may lose active work just as your computer loses active work when you restart it. It's important to keep in mind that many VMs restart automatically at a fixed time every day when not in use to save resources.

> That said, this is a secure means for executing a customer's scenario, and it works for any operating system and can even be configured to work for other architectures. Note that it does cost real money to use.

Reviewer 2: > Note that it does cost real money to use.

I think you're referring to VMs in the cloud.  It's also possible to create a local VM (for example using Hyper-V).

> That said, this is a secure means for executing a customer's scenario, and it works for any operating system and can even be configured to work for other architectures. Note that it does cost real money to use.

Author: ```suggestion
That said, this is a secure means for executing a customer's scenario, and it works for any operating system and can even be configured to work for other architectures. Note that using a VM in the cloud costs real money, though a local VM (such as using [Hyper-V](https://learn.microsoft.com/windows-server/virtualization/hyper-v/get-started/create-a-virtual-machine-in-hyper-v)) typically does not.
```

For Microsoft employees, [DevTestLab](https://ms.portal.azure.com/#browse/Microsoft.Compute%2FVirtualMachines) can help you create VMs.

### Read Code

> If none of the above methods work, it may be viable to read all customer code carefully to ensure that no part of it is risky or malicious.

Reviewer: It should be noted here that a basic editor should be prefered. Any IDE (VS, Raider etc.) or extensions (DevKit) that run design time build can cause unintended and unnoticed execution of untrusted code.

2. MSBuild SDKs. At the top of many .NET projects is a line that includes `Sdk="<SDK>"`. This automatically imports build logic associated with the specified SDK. If you are unfamiliar with the specified SDK, ensure it is secure before building any code including it.
3. MSBuild tasks. Within a `Target` block are MSBuild Tasks. Many of these, including `Exec` and `DownloadFile` execute logic that pulls code or build logic from the internet and may execute it. If you are unfamiliar with any MSBuild Task in a customer-provided project, ensure you fully understand its semantics before executing it (by building).
4. Typos. NuGet packages starting with Microsoft. and System. can only be pushed to nuget.org by Microsoft accounts, but Mircosoft and Sysstem are unregulated.
5. Malicious NuGet.Config files. Even if a PackageReference points to a package you're familiar with, perhaps even a Microsoft. or System. package, a NuGet.Config file can dictate searching in a different package source first that may not be secure and may include a malicious version of the package.

> More examples of patterns that can easily be exploited by malicious users can be found [here](https://aka.ms/msbuild-security-documentation)

Reviewer: Other things to watch for:
 * .rsp files injecting the custom extensibility
 * autoinclude files - Directory.Build.[props|target], .user project files - as additional logic can be injected via those
 * location where we open the projects from - as e.g. opening from 'Downloads' folder is supectible to injecting logic from unrelated auto-include files that can be placed to Downloads with relatively low privileges
 * attempts to write to locations outside of the root of current repo/solution (especially to locations like drive roots, msbuild install location, nuget cache, etc.) - this can inject logic to subsequent unrelated builds
 * Some additional tasks of interest: [internal link] https://microsoft.sharepoint.com/:w:/t/ManagedLanguagesIDE/EX0mgFi9TGJEkAKi3JYd7_EBZxyaXX1-tIze9Hbi2megLw?e=QUnpME - this however doesn't feel like it should be in scope of this doc (especially on public repo)

Author: These are good additions, but I don't want to overload this section with too many things to watch out for. I'd rather people following this option go to the security documentation you're working on and see a more complete set there 🙂

Reviewer: @Forgind - I realized I forgot to get back to you :-) The [MSBuild Security Best Practices](https://learn.microsoft.com/en-us/visualstudio/msbuild/msbuild-security-best-practices) external MS Learn doc is out there publicly available and the aka.ms link is redirecting to it (both has been done couple months back - I just forgot to mention it :))

Author: Yep, thanks! I got a notification when you changed the aka.ms link 🙂 Thanks for putting it together!

In addition to those more specific examples, it is important to follow the principle of least privilege. This means, for instance, that when on linux, you should only run a command as root if that's truly necessary to reproduce the customer's scenario. On Windows, you should only use an admin terminal if it's necessary.

## Response to Malicious Repros

If you discover that a customer provided a malicious project or solution, there are several steps. You should do all of these if possible.

1. Most importantly, do not keep any vestige of the repro on your machine. Do not follow their steps.
2. Add a note to the issue that it was malicious. Include this information in the title if possible.
3. Delete the contents of the repro steps. (On GitHub, it can still be found by looking at your edit.)

> 4. Report your finding to your manager and your security team.

Reviewer: https://aka.ms/reportitnow

To provide information to customers about collecting binlogs, direct them to:
https://aka.ms/msbuild/binlog

It is important to note, however, that binlogs often collect secrets as part of the information relevant to the build. A binlog of a simplified repro can work as well, but it may be difficult to create a simple repro. Customers are sometimes concerned (with good reason) about sharing binlogs of their real (or simplified) builds. If so, they can attempt to redact secrets from their binlog, either via the [MSBuild Structured Log Viewer](https://msbuildlog.com/#redaction) or with [binlogtool](https://www.nuget.org/packages/binlogtool). If that isn't sufficient, or if the reported issue does not involve MSBuild, another option may be more relevant.

> 

Author: I asked KirillOsenkov:
"Is it possible to craft a malicious binlog?"
and he said:
"I really don't think so"

The structured log viewer has an assumption of what the type is of everything it's going to read. If it sees something that's in the right form, it initializes it. If not, it throws an error. The types are fairly basic individually, so it isn't a viable attack vector, as far as we could tell.
